### DISCUSSION TITLE
Nx 16 Roadmap

### DISCUSSION THREAD

**Discussion by vsavkin** [nrwl/nx#12836] [General]
A few things happened in the last six months that changed our plans a bit:

1. We took stewardship of Lerna. 
2. The Nx community grew faster than we predicted.

Both are great things but they made us shift our priorities a bit. I believe Nx 15 is much better than we had planned but some of the things we wanted to do got delayed. 

## Some highlights from the last six months

We made changes to the Nx core such that it can be used to power the Lerna task orchestration. Even though the main goal was to make it compatible with Lerna (Lerna and Nx execution models aren’t identical, so Nx had to be generalized to handle both), it forced us to refactor the task orchestration in Nx. The result is a cleaner, easier-to-understand codebase. We fixed a ton of issues along the way. This also forced us to separate different parts of the Nx core cleanly. For instance, you can now run “npx nx graph” in any JS monorepo.

We generalized hashing. While it was always possible to write a custom hasher that would, for instance, ignore spec files when building the application, few did it. We made it declarative and concise–just two lines to define such a rule. When migrating to Nx 15, migration scripts will analyze your workspace to configure such rules for you. You will have a lot more cache hits when using Nx 15 than when using Nx 14.

More work was offloaded to the daemon, making running commands even faster. For instance, now the daemon keeps track of when you manually change files in “dist”, so Nx knows when they have to be restored from the cache and when they can be left as is. It’s also the daemon that keeps track of what projects you have in your workspace, helping you to avoid an expensive glob.

Because Nx has been around for six years, it has accrued some complexity where there is more than one way of doing certain things (e.g., you can use workspace.json or project.json). We have been updating workspaces to the new way of doing things and removing the legacy way, all done in a non-breaking fashion.

Because the Nx usage has grown so fast, we have many more issues and PRs. 

### Angular Plugin

The Angular Plugin for Nx has seen new features implemented, bugs fixed, and some cleanup regarding deprecated APIs in the last six months. We added support for Angular 14.1 and 14.2, allowing you to take full advantage of the latest features and bug fixes from Angular itself when you run `nx migrate latest`.

We have implemented much improved Module Federation support, as well as support for Standalone Components. 

We have also improved the automated migration from Angular CLI (learn more here: https://nx.dev/recipes/adopting-nx/migration-angular)

### Storybook

The past 6 months, we focused on making the Storybook integration much lighter. Our executors now invoke the native Storybook executors for Angular, and integrate tightly with the top level builder in other cases. This removed any inconsistencies between the native Storybook experience and the Storybook experience through Nx. 

We added generators for Vite, SWC, Next.js, Angular Standalone Components. We also added the ability to generate TypeScript configuration files.

### Nx Console

Nx Console has improved its architecture and introduced many new features. 

We created a language server for Nx, called the nxls. We use this as the core for Nx Console on Visual Studio code. Doing this first allows us to be one step closer to bringing Nx configuration autocomplete (and other features) to other IDEs and editors. 

For new features within Nx Console, we brought the project graph within the editor itself. This means that you can easily focus the graph on the project you’re working on and see what it could potentially affect. 

We can now show you if you’re connected to Nx Cloud, and offer a way to connect it for you automatically. 

We are also doing full affected e2e runs whenever a PR is opened. We download a full version of Visual Studio Code and run it headless on linux instances. We test workspaces that are Nx, Lerna, and even Angular based to ensure that all our cases are covered. 

### Docs

We restructured our documentation on nx.dev to more clearly separate task-oriented recipes from conceptual guides.  The introductory content is now focused on allowing you to start using a basic version of the feature as quickly as possible, then there are links to more detailed explanations.

— 

Read more about what we accomplished [here](https://blog.nrwl.io/whats-new-in-nx-15-7e14e1ff282d).

Now let’s see what is coming in the next 6 months.

### Nested projects and root-level scripts/tasks

We have been contemplating supporting root-level (aka workspace or projectless) scripts/tasks for some time, but we lacked an important core feature: an easy way to configure task inputs, which we have now. But after thinking about it for a bit more we realized that we could build a simpler and more general feature that will handle the root-level scripts use case + a few other high-priority use cases. 

Currently, projects in Nx cannot contain each other. If “packages/myproj” is the root of a given project, it means that no other project can be contained within that folder. If we remove this constraint, we will be able to create a root-level project in multi-project workspaces, so all root-level npm scripts would just work the way npm scripts work for any other project. They could be part of dependsOn). Caching would work etc.

This would also make Nx A LOT easier to adopt in non-monorepo workspaces, in cases where you have an app and a few axillary packages. 

### Lock file management

Currently, when computing a task hash we include the whole lock file into every task’s hash. This isn’t efficient. We have landed some initial support of analyzing lock files and only including what is relevant. A lot more is coming in the next 2 months.

We are also adding the ability to generate a pruned lock file that can be used by microservices.

### Negative configuration

Negative configuration is the idea where we can configure, say, the test target for 1000 projects in your workspace. So when you add Nx to your monorepo, the amount of configuration gets reduced (hence the name).

We have made some good progress in this area in the last six months. You can now define targetDefaults in nx.json. But more work is required to make it more complete.

### Workspace Watching

Monorepos can make some workflows more challenging. Say you have a monorepo and you want to watch Package A + everything it depends on such that when something changes, you rebuild the packages (or retest them). You could get the transitive closure of files, get the common ancestor, set up a file watcher etc, but it’s hard to do it well. Some of the things you need to handle: are excluding some fields from watching, handling removing or adding new edges in the graph, and coalescing notifications. Nx Daemon already does all of it! Now we can use this capability to write general purpose watchers, like this:

```
nx watch --projects=proj1 -- nx build proj1 # watch proj1 with all it deps, invoke nx build proj1 on changes.
```

### Plugins

In Nx 14 we introduced  some static analysis helping enhance nx-plugin quality in the form of an eslint rule. The first set of these checks ensures that collection files are valid, preventing an easy source of errors. We would like to expand these capabilities in Nx 16.

We plan to continue to enhance the developer experience for plugin authors. In particular, we would like to expand our existing plugin e2e setup to better handle workspaces that contain multiple plugins, or that make use of shared libraries. 

The combination of better static checks and better e2e will help increase the ease of maintaining a community plugin, but also help increase the overall quality and confidence in community plugins.

## JavaScript and React

In Nx 15, we added the esbuild bundler as an option for JS projects so it can be used in addition to normal tsc and rollup setup. For Nx 16, we want to add support for Vite bundler to better support various web frameworks such as React and Vue.

In addition to Vite, we will be adding support for server-side rendering (SSR) and server components in React. Although it is already possible to do SSR in React, either via a meta framework like Next.js, or hand-roll your own implementation, many teams find it challenging to set up SSR in their existing server. The new SSR generator will add a minimal setup to an existing Express or NestJS server, or generate a new server project with SSR enabled. In combination with React server components, we can improve performance by doing work such as data-fetching on the server side, and also reduce bundle size by skipping some client components altogether.

For JavaScript projects, we’re improving our workspace library support, so users can choose which workspace libraries to inline. This will allow non-buildable or non-published workspace libraries to be consumed in another library. We are also improving the esbuild plugin, and switching our Nx Console extension from webpack to esbuild.

### Angular 15 Support

We will add support for Angular 15 after it is released. We are working with the Angular team to make sure the migration process is painless. 

### Angular Universal (SSR) Support

Nx has never had any tools to help you install and generate Angular Universal to enable SSR for your Angular applications as we have focused on other priorities but we feel the time is now right to add support in the form of generators to help you both add SSR support to an existing Angular application as well as generate an SSR-ready Angular application from scratch. \
 \
As the official support for Angular Universal is scattered across many packages in the Angular ecosystem, we feel this simplified approach will be much appreciated for all Nx users who are looking to add Angular Universal to their applications.

As part of this, we will add a new builder to the Angular Plugin that will allow you to use a custom webpack configuration to build your server-side application which many may find useful to allow them to apply additional Webpack Plugins they may need.

### Module Federation Support for SSR

Expanding on our current Module Federation support, which is currently limited to client-side applications, we will offer support for taking advantage of the benefits of Module Federation for SSR Angular applications that use Angular Universal.

### Improvements to NgRx Support

Our NgRx generators have fallen by the wayside due to priorities in other areas. For Nx 16, we will be carefully reviewing the support we offer for generating NgRx-related code and will aim to offer much improved support and generators to aid you in rapid development of your State Management for your applications.

We will also make changes to the code that is generated to keep it in line with current improvements made by the NgRx team, such as Standalone APIs and `createActionGroup`.

### Angular CLI to Nx workspace migration improvements

Migrating Angular CLI workspaces to Nx in an automated way was greatly improved by handling workspaces with multiple projects and more configuration scenarios than before, but we’re not done yet. The migration currently only handles configured targets using Angular CLI official builders. In the coming months, we’ll work on adding support for other popular builders widely used by the community to help migrate workspaces using them.

#### Storybook v7 support and migrators

Storybook v7 is almost here. It has a number of breaking changes, so our goal is to make sure the necessary migrators exist to ensure the updating experience is as easy as possible. We want to enable Nx users to update to v7 painlessly. 

#### Keep Storybook integration complete and up to date

We are continuously working closely with the Storybook team to make sure the integration of our tools is complete, by being proactive for any breaking changes, and attending to bugs as soon as possible.

#### Tools for publishing and CI

Another goal for the next months is to create some tooling that will help with the publishing strategies that we suggest. Working with the Storybook team, we also aim to fix our parallel e2e test execution.

#### Cypress 

Writing large and involved Cypress tests oftens leads to extracting out common functionality into a local package. While helpful, these extracted pieces of functionality can be dense and hard to maintain. Our goal is to add a Cypress library support to help streamline and setup best practices for authoring and maintaining your external Cypress logic

#### Jest

Setting up Jest is one thing Nx already does for you, but after that it can be tricky to maintain any non-trivial project's Jest configuration with changing dependencies. Our goal is to make this simple by collecting your project dependencies and keeping Jest transformations list up to date automatically for you.

### Visual Studio Code

We’re planning on adding more ways to see your Nx Cloud visualizations within your workspace. A few things that are planned will be:

* Easily find the latest run link for your PR
* Add more visualizations to see when your cloud runs begin and start

And of course any new features that are requested by the community where we see fit.

### Nx Language Server

The language server is going to be an integral part of expanding Nx Console to other editors and IDEs. It’s also going to be more fully integrated with Visual Studio Code. So we’re going to continue to improve on it by helping other community extensions integrate with it, and add new features that are related to helping you configure your workspace better.

**Comment by gioragutt** [comment]
How about libraries publishing scenarios? Any plans on that?

**Reply by datumgeek** [reply]
UPDATE: code published to Comcast Open Source repo: [https://github.com/Comcast/Porrtal](https://github.com/Comcast/Porrtal )
UPDATE: new finding: problem might have been caused by the directory structure I chose for my packages.
UPDATE: with the corrected path structure, publishing works beautifully for both angular and react.  It is automatically generating the peerDependencies in the built package.json.  I am 100% unblocked...

This choice seems to not work when building mylib which depends on mylib-two:

| package | folder | platform |
| --------- | ------- | --------- |
| @myscope/r-mylib | libs/myscope/r/mylib | react |
| @myscope/a-mylib | libs/myscope/a/mylib | angular |
| @myscope/r-mylib-two | libs/myscope/r/mylib-two | react |
| @myscope/a-mylib-two | libs/myscope/a/mylib-two | angular |

But this choice does seems to work (so far - I'm continuing testing)

| package | folder | platform |
| --------- | ------- | --------- |
| @myscope/r-mylib | libs/myscope/r-mylib | react |
| @myscope/a-mylib | libs/myscope/a-mylib | angular |
| @myscope/r-mylib-two | libs/myscope/r-mylib-two | react |
| @myscope/a-mylib-two | libs/myscope/a-mylib-two | angular |

[Porrtal Nx Libraries](https://github.com/Comcast/Porrtal/tree/main/libs/porrtal)
[Porrtal NPM Packages](https://www.npmjs.com/search?q=%40porrtal)

As a developer of open source npm packages, I want Nx to be able to build, test and serve my projects locally and then be able to build and publish them to NPM.

Nx is amazing and the increase in usage is astounding.  The use case above is really my number one, super important feature I need from Nx.  It makes me a little crazy because it is blocking me from publishing an exciting open source project I've been building for 6 months.  Please, please, please, please look at this request :)

I'm building both Angular and React libraries that are to be published to NPM packages.

The build errors for both Angular and React libraries occur when on publishable library references another publishable library.

**Comment by luxaritas** [comment]
"Workspace watching" winds up being tied to #5570, which I've explored various solutions to in the past. I was involved with a discussion with James Henry in the community Slack where he mentioned that this was the intended solution to this problem. While I think it is an elegant solution, there are two open questions I still have:
1) Has there been/is there plans for benchmarking the cost of eg starting multiple build processes on each change as opposed to keeping per-project watch scripts running? It would certainly mean that you don't have the overhead of keeping all of those processes running, however I'm wondering if it would ultimately increase time-to-completion (particularly if you have a long dependency chain, you add the latency of starting the build for each dependency plus doing a full build, vs in some cases being able to just rebuild one file if you don't need to bundle).
2) As far as support for non-exiting dependency tasks, has there been any thought in managing eg e2e scripts where it's strictly necessary to start a dev server in some other package? Would the solution just be `concurrently nx serve x && <actually run tests>`? I could imagine this could create issues if multiple packages rely on the same server to be up, and just generally the fact that nx would be internally running nx without them "knowing about each other".

**Reply by wrslatz** [reply]
For point 2, https://github.com/nrwl/nx/issues/11827#issuecomment-1292605350 and the linked issues are related

**Comment by Waterstraal** [comment]
I'm curious, why not drop the NgRx generator altogether? It doesn't feel like a core Nx feature to maintain imho. 

Alternatively, you could add generators for rx-angular/state, ngxs, elf, akita to cover all the major state management libs :).

**Comment by sir-captainmorgan21** [comment]
Lack of SSR support for Module Federation is the one thing keeping us from using Module Federation. Are there any working solution that anyone know of in the meantime until support is baked into NX?

**Reply by sir-captainmorgan21** [reply]
For those who are also looking for solutions. The module federation team has a ton of examples: https://github.com/module-federation/module-federation-examples/tree/master/angular-universal-ssr Going to try this one out

**Reply by jasonrichdarmawan** [reply]
Is Incremental Static Rendering supported? https://github.com/rx-angular/rx-angular

**Comment by asalahjawava** [comment]
Is there any plan to support Module Federation for SSR React with NextJS?

**Comment by mcblum** [comment]
We're evaluating whether to move to ViteJS right now so we very-much appreciate the consideration in Nx 16!

**Reply by gioragutt** [reply]
Just saying, I was able to use a nrwl/react:lib generated library with an app generated with vite create, it just worked out of the box (almost entirely). You might not necessarily have to wait for an official NX plugin.

**Reply by gioragutt** [reply]
```ts
import { defineConfig } from 'vite'
import react from '@vitejs/plugin-react'
import tsconfigPaths from 'vite-tsconfig-paths'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [react(), tsconfigPaths()]
})
```

All I had to do was add `vite-tsconfig-paths`.

**Reply by mcblum** [reply]
That's cool! Currently our monorepo is 100% Webpack, so it would be more of a conversion. I used the @nrwl/react app generator, not `vite create`.

**Reply by mcblum** [reply]
Also, I don't know much about ViteJS, so perhaps my comment there makes no sense :)

**Reply by gioragutt** [reply]
No no you're absolutely right. But the nice thing is that you can totally just POC adding another app, have it be one created with `vite create` (yes yes, package.json, node_modules and all), and see what would be the cost of using it.

That means:

1. Try to import libraries from the rest of the app
2. Try to use resources that are handled by webpack (things like SVGs or whatever you have loaders for) and see if it's handled by vite, or whether there are plugins for those things.

IMO, if you have time, just play with it. In the worst case, you've seen how freakishly fast vite is, and you've gained some experience with it.

**Reply by nvonbenken** [reply]
@gioragutt Any chance you've tried a module federation approach with NX, Vite, and [vite-plugin-federation](https://www.npmjs.com/package/@originjs/vite-plugin-federation)?

**Reply by gioragutt** [reply]
@nvonbenken I wish I had a use case to try that, I really wish 😅 sorry but I did not

**Reply by nvonbenken** [reply]
I guess its time for me to give it a go then!

**Reply by gioragutt** [reply]
Good luck then! I'd love to see the prototype :)

**Reply by Jordan-Hall** [reply]
If you want to use vite for now then look at https://nxext.dev/ nx 16 will come with or we will provide migrations for nx own vite.

**Comment by c0ncentus** [comment]
the main drawback of my NX's experience: 

I come to CRA project (package oriented) so i really don't want to touch a single line of webpack things ...
but in my workspace ts.conf.base.json's lib dont apply compiler config so I have 3 errors and cannot use my custom package because of compiler issue ...

CRA complex experience (sharaing assets/ sharing UI component)  isnt good for me ... 

hope thats change ...

**Comment by moatorres** [comment]
Just popping here to tell that I can't express how much I love Nx. Absolute game changer. @vsavkin, you are a superstar! 

I'd love to help the team with ~~anything~~ something. I'm not as prolific as I think you guys are, but maybe I can fix typos or tidy things up? 😄 

One of the most interesting things to me is that whenever navigating/reading through the codebase, I stumble upon some "rushed" lines, as in "not perfectly tidy or optimal, but working code", and it reminds me of how important it is to build things and not worry too much with tidiness (OCD-ing?) or "style", but to deliver value to the community. Nx inspires me in that sense of "build fast and don't worry too much about it". 

Keep up with the good work! 🥳

**Comment by lumiiartan** [comment]
Does anyone know if Vue will be supported soon for Cypress component testing in Nx, considering that Nx currently supports Angular and React?

**Reply by JFGHT** [reply]
> we want to add support for Vite bundler to better support various web frameworks such as React and Vue

Strange that this has not gathered enough attention.

**Reply by gioragutt** [reply]
Probably because just adding vite support, regardless of which framework you're aiming to cater, was enough 🤷🏻‍♂️

**Comment by airtonix** [comment]
What's the plan for completing the storybook composition story.

Right now, from my point of view it looks like a developer has to walk off-pist to build and deploy a monorepo that has composed storybook instances.

How are people automating deployments of this now? 

I feel like there's some interaction I'm missing between `nx affected`, pull-request previews and not having to completly regenerate all storybook stories every single build.


the examples given so far only talk about the host wrapper with a `refs` that contains `http://localhost:....`, but this only works in local development. Surely we'd want something for pull request previews that land on a auto generated subdomain?


If i have to rebuild everything on every commit, i don't see the point of composed storybook setup.

**Comment by Ntnelbaba** [comment]
How can we know when this version will be released?

**Reply by ild0tt0re** [reply]
@Ntnelbaba released -> https://blog.nrwl.io/nx-16-is-here-69584ec87053

**Reply by Ntnelbaba** [reply]
@ild0tt0re Thanks!

Just for the core team to know - when there is a roadmap for a version, an estimation of a release month will be very helpful.

**Comment by phihochzwei** [comment]
Will there be support for NG16 ?

**Comment by nhhockeyplayer** [comment]
Nx 16 looking fab, NxConsole heavy artillery and much more for 2023 
Thanks Nx team for delivering !!!

**Comment by sgtsquiggs** [comment]
Fingers crossed for Vite Module Federation support (https://github.com/nrwl/nx/issues/13354). NX support for MF w/ webpack works great but a lot of us are using Vite!

**Comment by alfaproject** [comment]
Nx 17 just dropped, so I guess this either should be unpinned, or a new one made for next version? It would be nice to know what your plans are in general strokes

**Comment by Nesho-Developer** [comment]
Is there any plan to support esbuild with Angular Module Federation?
